File: 99-testing-the-domain/DomainTestDemo.java
/*
 * Ngôn ngữ: Java
 * Công dụng: Test mô hình miền — không khung phần mềm, không CSDL, không mock. Bài cho nổ
 * ba con bug của cách test phổ biến: test bám vào CÁCH LÀM nên refactor không đổi hành vi
 * vẫn làm đỏ 3 test; test dựng dữ liệu bằng 12 dòng nên không ai đọc được nó kiểm gì; và
 * bản giả nói dối so với bản thật mà không test nào phát hiện.
 * Tại sao cần học: cả tầng này đã dựng một mô hình miền không phụ thuộc hạ tầng (bài 98).
 * Bài này thu hoạch: khi miền không biết CSDL, test của nó chỉ còn là HÀM + ASSERT — đúng
 * như mọi file self-check trong series này. Không phải "chúng ta lười không dùng JUnit";
 * đó là bằng chứng rằng mô hình miền đã tách sạch. Nếu test miền của bạn cần một khung
 * phần mềm để chạy, thì thứ bạn đang test không phải miền.
 */
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;

public class DomainTestDemo {

    // =====================================================================
    // MIỀN — đối tượng được test
    // =====================================================================
    record DongHang(String sanPham, long donGia, int soLuong) {
        long thanhTien() { return donGia * soLuong; }
    }

    static final class DonHang {
        static final long HAN_MUC = 50_000_000L;
        private final String ma;
        private final List<DongHang> cacDong = new ArrayList<>();
        int soLanGoiKho = 0;                       // để đo ở phần 1

        DonHang(String ma) { this.ma = ma; }

        void themDong(String sp, long donGia, int sl, KhoHang kho) {
            if (tongTien() + donGia * sl > HAN_MUC)
                throw new IllegalStateException("đơn vượt hạn mức");
            kho.giuCho(sp, sl);
            soLanGoiKho++;
            cacDong.add(new DongHang(sp, donGia, sl));
        }

        /** Phiên bản gộp: giữ chỗ MỘT lần cho nhiều dòng. Hành vi ngoài KHÔNG đổi. */
        void themNhieuDong(List<DongHang> ds, KhoHang kho) {
            long them = ds.stream().mapToLong(DongHang::thanhTien).sum();
            if (tongTien() + them > HAN_MUC) throw new IllegalStateException("đơn vượt hạn mức");
            Map<String, Integer> gop = new LinkedHashMap<>();
            for (DongHang d : ds) gop.merge(d.sanPham(), d.soLuong(), Integer::sum);
            kho.giuChoNhieu(gop);                  // MỘT lượt gọi thay vì n lượt
            soLanGoiKho++;
            cacDong.addAll(ds);
        }

        long tongTien() { return cacDong.stream().mapToLong(DongHang::thanhTien).sum(); }
        int soDong() { return cacDong.size(); }
        String ma() { return ma; }

        /** Chia đều tiền đơn cho n người — bất biến ở phần 3 (bài 90). */
        List<Long> chiaDeu(int n) {
            long tong = tongTien(), moi = tong / n, du = tong % n;
            List<Long> ra = new ArrayList<>();
            for (int i = 0; i < n; i++) ra.add(moi + (i < du ? 1 : 0));
            return ra;
        }
    }

    interface KhoHang {
        void giuCho(String sanPham, int soLuong);
        void giuChoNhieu(Map<String, Integer> gop);
    }

    /** Bản GIẢ — dùng cho test. Nó là một cài đặt thật, không phải mock. */
    static final class KhoGia implements KhoHang {
        final Map<String, Integer> daGiu = new LinkedHashMap<>();
        int soLuotGoi = 0;
        @Override public void giuCho(String sp, int sl) { soLuotGoi++; daGiu.merge(sp, sl, Integer::sum); }
        @Override public void giuChoNhieu(Map<String, Integer> gop) {
            soLuotGoi++;
            gop.forEach((k, v) -> daGiu.merge(k, v, Integer::sum));
        }
    }

    /** Bản GIẢ NÓI DỐI — cho phần 4. Nó quên mất luật "không giữ chỗ số lượng âm". */
    static final class KhoGiaNoiDoi implements KhoHang {
        final Map<String, Integer> daGiu = new LinkedHashMap<>();
        @Override public void giuCho(String sp, int sl) { daGiu.merge(sp, sl, Integer::sum); }
        @Override public void giuChoNhieu(Map<String, Integer> gop) { daGiu.putAll(gop); }
    }

    /** Bản "thật" — có thêm một luật mà bản giả nói dối không có. */
    static final class KhoThat implements KhoHang {
        final Map<String, Integer> daGiu = new LinkedHashMap<>();
        @Override public void giuCho(String sp, int sl) {
            if (sl <= 0) throw new IllegalArgumentException("số lượng giữ chỗ phải dương");
            daGiu.merge(sp, sl, Integer::sum);
        }
        @Override public void giuChoNhieu(Map<String, Integer> gop) {
            gop.forEach(this::giuCho);
        }
    }

    // =====================================================================
    // BỘ DỰNG DỮ LIỆU TEST — phần 2
    // =====================================================================
    static final class DonHangBuilder {
        private String ma = "DH-MAU";
        private final List<DongHang> dong = new ArrayList<>();
        DonHangBuilder ma(String m) { this.ma = m; return this; }
        DonHangBuilder voiDong(String sp, long gia, int sl) { dong.add(new DongHang(sp, gia, sl)); return this; }
        DonHang dung(KhoHang kho) {
            DonHang d = new DonHang(ma);
            if (!dong.isEmpty()) d.themNhieuDong(dong, kho);
            return d;
        }
    }

    // ---- Self-check ----
    private static void check(boolean ok, String msg) {
        if (!ok) throw new AssertionError(msg);
    }

    public static void main(String[] args) {
        // ---- 1. CON BUG: test bám vào CÁCH LÀM, không bám vào HÀNH VI ----
        // Ba "test" dưới đây kiểm SỐ LƯỢT GỌI kho — thứ mà mock framework khuyến khích
        // (`verify(kho, times(3)).giuCho(...)`).
        KhoGia khoA = new KhoGia();
        DonHang donA = new DonHang("DH-01");
        donA.themDong("laptop", 1_000_000L, 1, khoA);
        donA.themDong("chuột", 200_000L, 2, khoA);
        donA.themDong("bàn phím", 300_000L, 1, khoA);
        check(khoA.soLuotGoi == 3, "test-theo-cách-làm: 'kho phải được gọi ĐÚNG 3 lần'");

        // Hôm nay ai đó gộp ba lượt gọi thành một để giảm tải cho kho. HÀNH VI NGOÀI KHÔNG
        // ĐỔI: cùng số dòng, cùng tổng tiền, cùng số lượng được giữ.
        KhoGia khoB = new KhoGia();
        DonHang donB = new DonHang("DH-01");
        donB.themNhieuDong(List.of(new DongHang("laptop", 1_000_000L, 1),
                                   new DongHang("chuột", 200_000L, 2),
                                   new DongHang("bàn phím", 300_000L, 1)), khoB);

        check(khoB.soLuotGoi == 1, "test-theo-cách-làm giờ ĐỎ: 1 ≠ 3");
        check(donA.tongTien() == donB.tongTien(), "nhưng tổng tiền GIỐNG HỆT");
        check(donA.soDong() == donB.soDong(), "cùng số dòng");
        check(khoA.daGiu.equals(khoB.daGiu), "và kho giữ chỗ ĐÚNG NHƯ NHAU");
        // Ba dòng cuối là test-theo-HÀNH-VI, và cả ba vẫn xanh. Đó là toàn bộ khác biệt:
        //   - Test hành vi hỏng khi NGHIỆP VỤ sai      -> tín hiệu THẬT.
        //   - Test cách làm hỏng khi CODE ĐỔI          -> tín hiệu GIẢ.
        // Bộ test đầy tín hiệu giả là bộ test bị tắt sau ba tháng. Quy tắc: đừng kiểm "đã
        // gọi hàm nào bao nhiêu lần"; kiểm "kết quả có đúng không, trạng thái có đúng không".
        //
        // Ngoại lệ hợp lệ duy nhất: khi VIỆC GỌI CHÍNH LÀ hành vi cần kiểm — "đã gửi đúng
        // một email cho khách" (bài 84 phần 4). Lúc đó số lượt gọi là nghiệp vụ, không phải
        // chi tiết cài đặt.

        // ---- 2. BỘ DỰNG DỮ LIỆU: test đọc lên phải nói được nó kiểm gì ----
        // Không có bộ dựng, mỗi test mở đầu bằng 10 dòng nhiễu và người đọc không thấy
        // được ĐIỀU GÌ trong dữ liệu là quan trọng với test này.
        KhoGia kho2 = new KhoGia();
        DonHang ganHanMuc = new DonHangBuilder()
                .ma("DH-02")
                .voiDong("máy chủ", 49_000_000L, 1)      // <- chi tiết DUY NHẤT quan trọng
                .dung(kho2);
        boolean vuotHanMuc = false;
        try { ganHanMuc.themDong("laptop", 2_000_000L, 1, kho2); }
        catch (IllegalStateException e) { vuotHanMuc = true; }
        check(vuotHanMuc, "đơn 49 triệu + 2 triệu -> vượt hạn mức 50 triệu");
        // Đọc ba dòng trên là biết ngay test kiểm gì. Bộ dựng có giá trị mặc định cho MỌI
        // thứ, và test chỉ nói ra thứ nó QUAN TÂM — đó là toàn bộ mục đích của nó.

        // ---- 3. BẤT BIẾN: kiểm với NGHÌN đầu vào, không phải ba ----
        // Test theo ví dụ trả lời "với đầu vào này thì sao". Test theo bất biến trả lời
        // "với MỌI đầu vào thì điều gì luôn đúng" — và nó bắt được những ca mà không ai
        // nghĩ ra để viết ví dụ.
        Random rnd = new Random(42);                 // hạt giống CỐ ĐỊNH -> tái hiện được
        int soCaChay = 0;
        for (int i = 0; i < 1000; i++) {
            KhoGia k = new KhoGia();
            long gia = 1 + rnd.nextInt(1_000_000);
            int sl = 1 + rnd.nextInt(40);
            int nguoi = 1 + rnd.nextInt(9);
            DonHang d = new DonHangBuilder().voiDong("x", gia, sl).dung(k);
            List<Long> phan = d.chiaDeu(nguoi);
            long tongPhan = phan.stream().mapToLong(Long::longValue).sum();
            check(tongPhan == d.tongTien(), "BẤT BIẾN: tổng các phần = tổng ban đầu");
            check(phan.size() == nguoi, "BẤT BIẾN: đúng số phần được yêu cầu");
            long max = phan.stream().mapToLong(Long::longValue).max().orElse(0);
            long min = phan.stream().mapToLong(Long::longValue).min().orElse(0);
            check(max - min <= 1, "BẤT BIẾN: chênh lệch giữa các phần không quá 1 đơn vị");
            soCaChay++;
        }
        check(soCaChay == 1000, "1.000 ca sinh ngẫu nhiên, 3 bất biến, 0 dòng dữ liệu gõ tay");
        // Ba bất biến đó là bài 90 phần 4 viết dưới dạng test. Chúng đắt hơn ví dụ ở chỗ
        // phải NGHĨ RA được, và rẻ hơn ở chỗ không phải bảo trì bảng dữ liệu.
        // Hạt giống cố định là bắt buộc: một test đỏ ngẫu nhiên mà không tái hiện được thì
        // vô dụng — và tệ hơn, nó sẽ bị đánh dấu "bỏ qua".

        // ---- 4. CON BUG: BẢN GIẢ NÓI DỐI ----
        // Bản giả không có luật "số lượng giữ chỗ phải dương". Test dùng nó thì xanh:
        DonHang donC = new DonHang("DH-03");
        KhoGiaNoiDoi noiDoi = new KhoGiaNoiDoi();
        donC.themDong("laptop", 1_000_000L, -5, noiDoi);        // số lượng ÂM
        check(noiDoi.daGiu.get("laptop") == -5, "bản giả nhận số lượng âm — test XANH");

        // ...và bản thật thì nổ, trên production.
        boolean banThatNo = false;
        try { new DonHang("DH-04").themDong("laptop", 1_000_000L, -5, new KhoThat()); }
        catch (IllegalArgumentException e) { banThatNo = true; }
        check(banThatNo, "bản thật ném ngoại lệ — bug đi thẳng ra production");
        // Cách chữa là BỘ KIỂM TRA HỢP ĐỒNG (bài 68): một bộ test viết một lần, chạy trên
        // MỌI cài đặt của cổng. Bản giả nào không qua được thì không được dùng.
        List<KhoHang> moiCaiDat = List.of(new KhoGia(), new KhoThat());
        int soCaiDatQua = 0;
        for (KhoHang k : moiCaiDat) {
            boolean chan = false;
            try { k.giuCho("x", -1); } catch (IllegalArgumentException e) { chan = true; }
            if (chan) soCaiDatQua++;
        }
        check(soCaiDatQua == 1, "chạy hợp đồng trên 2 cài đặt -> lộ ra ngay cái nào nói dối");
        // (Ở đây `KhoGia` cũng trượt — đúng như dự định: bộ kiểm tra hợp đồng vừa chỉ ra
        //  rằng bản giả trong bộ test này CẦN được sửa cho khớp với bản thật.)

        // ---- 5. TEST MIỀN KHÔNG CẦN KHUNG PHẦN MỀM ----
        // Toàn bộ file này — và mọi file self-check trong series — là hàm + `assert`. Không
        // JUnit, không mock, không `@SpringBootTest`, 0 mili-giây khởi động.
        //
        // Đó không phải "lười không dùng JUnit". Nó là BẰNG CHỨNG rằng mô hình miền đã
        // tách sạch khỏi hạ tầng (bài 98). Phép thử ngược lại cũng đúng và rất hữu ích:
        //
        //   NẾU TEST MIỀN CỦA BẠN CẦN MỘT KHUNG PHẦN MỀM ĐỂ CHẠY,
        //   THÌ THỨ BẠN ĐANG TEST KHÔNG PHẢI MIỀN.
        //
        // Trong dự án thật vẫn nên dùng JUnit — vì nó cho báo cáo, chạy song song, tích hợp
        // CI. Nhưng test miền phải chạy được KHÔNG CẦN nó, và điều đó phải đúng ngay từ đầu.
        check(donA.ma().equals("DH-01"), "một hàm, một assert, không hạ tầng nào");

        // ---- 6. CÁI GÌ KHÔNG NÊN TEST ----
        //   - Getter/setter thuần: `check(d.ma().equals("DH-01"))` không kiểm được luật nào.
        //   - Thư viện và khung phần mềm: `HashMap` đã được test rồi.
        //   - Cài đặt riêng tư: nếu phải đổi `private` thành `public` để test, thì test đó
        //     đang bám vào cách làm (phần 1) — hãy test qua cửa chính.
        //   - Bộ nối hạ tầng bằng test miền: bộ nối cần test tích hợp riêng, ít và chậm.
        // Ngược lại, thứ ĐÁNG test nhất là những chỗ có `if` mang nghĩa nghiệp vụ: hạn
        // mức, chuyển trạng thái (bài 89), luật giá, luật chia tiền.
        check(DonHang.HAN_MUC == 50_000_000L, "luật nghiệp vụ có `if` -> đáng test");

        // ---- 7. ĐẶT TÊN TEST NHƯ MỘT CÂU NGHIỆP VỤ ----
        // Ở series này, "tên test" là chuỗi thông báo trong `check(...)`. Nguyên tắc giống
        // hệt với `@Test void ...`:
        //   TỆ : "test1", "testThemDong", "kiểm tra hàm themDong"
        //   TỐT: "đơn 49 triệu + 2 triệu -> vượt hạn mức 50 triệu"
        // Tên tốt nói ĐIỀU KIỆN và KẾT QUẢ MONG ĐỢI, bằng từ ngữ nghiệp vụ (bài 81). Khi
        // test đỏ lúc 2 giờ sáng, dòng chữ đó là toàn bộ thứ người trực có.
        check(true, "tên test là tài liệu duy nhất không bao giờ lỗi thời");

        System.out.println("OK");
    }
}

File: 99-testing-the-domain/DomainTestDemo.cpp
/*
 * Ngôn ngữ: C++
 * Công dụng: Bản C++ của cùng bài học — test mô hình miền không khung phần mềm, không
 * CSDL, không mock. Ba con bug: test bám vào CÁCH LÀM; test dựng dữ liệu quá nhiễu; và
 * bản giả nói dối so với bản thật.
 * Tại sao cần học: C++ có một loại test mà Java và Python không có — test chạy LÚC BIÊN
 * DỊCH. Với `constexpr` + `static_assert`, một bất biến của miền được kiểm trước khi
 * chương trình tồn tại, tốn 0 mili-giây lúc chạy, và code sai thì KHÔNG TẠO RA ĐƯỢC file
 * thực thi. Đó là vòng phản hồi nhanh nhất có thể: không phải "test đỏ sau 3 giây" mà là
 * "build đỏ ngay trong trình soạn thảo". Bài chỉ ra phần nào của miền đưa được vào đó, và
 * phần nào thì không.
 */
#include <iostream>
#include <map>
#include <random>
#include <stdexcept>
#include <string>
#include <vector>
#include <cstdlib>

// =====================================================================
// MIEN — doi tuong duoc test
// =====================================================================
struct DongHang {
    std::string sanPham;
    long donGia;
    int soLuong;
    constexpr long thanhTien() const { return donGia * soLuong; }
};

class KhoHang {
public:
    virtual ~KhoHang() = default;
    virtual void giuCho(const std::string& sp, int sl) = 0;
    virtual void giuChoNhieu(const std::map<std::string, int>& gop) = 0;
};

class DonHang {
public:
    static constexpr long HAN_MUC = 50000000L;

    explicit DonHang(std::string ma) : ma_(std::move(ma)) {}

    void themDong(const std::string& sp, long donGia, int sl, KhoHang& kho) {
        if (tongTien() + donGia * sl > HAN_MUC) throw std::logic_error("don vuot han muc");
        kho.giuCho(sp, sl);
        cacDong_.push_back(DongHang{sp, donGia, sl});
    }

    // Phien ban gop: giu cho MOT lan cho nhieu dong. Hanh vi ngoai KHONG doi.
    void themNhieuDong(const std::vector<DongHang>& ds, KhoHang& kho) {
        long them = 0;
        for (const auto& d : ds) them += d.thanhTien();
        if (tongTien() + them > HAN_MUC) throw std::logic_error("don vuot han muc");
        std::map<std::string, int> gop;
        for (const auto& d : ds) gop[d.sanPham] += d.soLuong;
        kho.giuChoNhieu(gop);                          // MOT luot goi thay vi n luot
        cacDong_.insert(cacDong_.end(), ds.begin(), ds.end());
    }

    long tongTien() const {
        long t = 0;
        for (const auto& d : cacDong_) t += d.thanhTien();
        return t;
    }
    std::size_t soDong() const { return cacDong_.size(); }
    const std::string& ma() const { return ma_; }

private:
    std::string ma_;
    std::vector<DongHang> cacDong_;
};

// Chia deu tien cho n phan — bat bien o phan 3 (bai 90). `constexpr` de dung o phan 5.
constexpr long phanThuI(long tong, int n, int i) {
    return tong / n + (i < tong % n ? 1 : 0);
}

// Ban GIA — mot cai dat that, khong phai mock.
class KhoGia : public KhoHang {
public:
    void giuCho(const std::string& sp, int sl) override { ++soLuotGoi; daGiu[sp] += sl; }
    void giuChoNhieu(const std::map<std::string, int>& gop) override {
        ++soLuotGoi;
        for (const auto& [k, v] : gop) daGiu[k] += v;
    }
    std::map<std::string, int> daGiu;
    int soLuotGoi = 0;
};

// Ban "that" — co them mot luat ma ban gia khong co.
class KhoThat : public KhoHang {
public:
    void giuCho(const std::string& sp, int sl) override {
        if (sl <= 0) throw std::invalid_argument("so luong giu cho phai duong");
        daGiu[sp] += sl;
    }
    void giuChoNhieu(const std::map<std::string, int>& gop) override {
        for (const auto& [k, v] : gop) giuCho(k, v);
    }
    std::map<std::string, int> daGiu;
};

// BO DUNG DU LIEU TEST — phan 2
class DonHangBuilder {
public:
    DonHangBuilder& ma(std::string m) { ma_ = std::move(m); return *this; }
    DonHangBuilder& voiDong(std::string sp, long gia, int sl) {
        dong_.push_back(DongHang{std::move(sp), gia, sl});
        return *this;
    }
    DonHang dung(KhoHang& kho) {
        DonHang d(ma_);
        if (!dong_.empty()) d.themNhieuDong(dong_, kho);
        return d;
    }
private:
    std::string ma_ = "DH-MAU";
    std::vector<DongHang> dong_;
};

// =====================================================================
// PHAN 5 — TEST CHAY LUC BIEN DICH
// =====================================================================
constexpr long tongPhan(long tong, int n) {
    long s = 0;
    for (int i = 0; i < n; ++i) s += phanThuI(tong, n, i);
    return s;
}
constexpr long chenhLech(long tong, int n) {
    long lon = phanThuI(tong, n, 0), nho = phanThuI(tong, n, n - 1);
    return lon - nho;
}
// Ba bat bien cua bai 90, kiem TRUOC khi chuong trinh ton tai:
static_assert(tongPhan(100, 3) == 100, "tong cac phan phai bang tong ban dau");
static_assert(tongPhan(1, 7) == 1, "ke ca khi tong nho hon so phan");
static_assert(tongPhan(999999937, 9) == 999999937, "va voi so nguyen to lon");
static_assert(chenhLech(100, 3) <= 1, "chenh lech giua cac phan khong qua 1 don vi");
static_assert(chenhLech(1, 7) <= 1, "ke ca truong hop bien");

// ---- Self-check ----
static void check(bool ok, const std::string& msg) {
    if (!ok) {
        std::cerr << "FAIL: " << msg << "\n";
        std::exit(1);
    }
}

int main() {
    // ---- 1. CON BUG: test bam vao CACH LAM, khong bam vao HANH VI ----
    KhoGia khoA;
    DonHang donA("DH-01");
    donA.themDong("laptop", 1000000L, 1, khoA);
    donA.themDong("chuot", 200000L, 2, khoA);
    donA.themDong("ban phim", 300000L, 1, khoA);
    check(khoA.soLuotGoi == 3, "test-theo-cach-lam: 'kho phai duoc goi DUNG 3 lan'");

    // Hom nay ai do gop ba luot goi thanh mot. HANH VI NGOAI KHONG DOI.
    KhoGia khoB;
    DonHang donB("DH-01");
    donB.themNhieuDong({{"laptop", 1000000L, 1}, {"chuot", 200000L, 2}, {"ban phim", 300000L, 1}}, khoB);

    check(khoB.soLuotGoi == 1, "test-theo-cach-lam gio DO: 1 != 3");
    check(donA.tongTien() == donB.tongTien(), "nhung tong tien GIONG HET");
    check(donA.soDong() == donB.soDong(), "cung so dong");
    check(khoA.daGiu == khoB.daGiu, "va kho giu cho DUNG NHU NHAU");
    // Ba dong cuoi la test-theo-HANH-VI, va ca ba van xanh. Khac biet:
    //   - Test hanh vi hong khi NGHIEP VU sai  -> tin hieu THAT.
    //   - Test cach lam hong khi CODE DOI      -> tin hieu GIA.
    // Bo test day tin hieu gia la bo test bi tat sau ba thang.
    //
    // Ngoai le hop le duy nhat: khi VIEC GOI CHINH LA hanh vi can kiem — "da gui dung mot
    // email cho khach" (bai 84 phan 4).

    // ---- 2. BO DUNG DU LIEU: test doc len phai noi duoc no kiem gi ----
    KhoGia kho2;
    DonHang ganHanMuc = DonHangBuilder()
                            .ma("DH-02")
                            .voiDong("may chu", 49000000L, 1)   // <- chi tiet DUY NHAT quan trong
                            .dung(kho2);
    bool vuotHanMuc = false;
    try { ganHanMuc.themDong("laptop", 2000000L, 1, kho2); }
    catch (const std::logic_error&) { vuotHanMuc = true; }
    check(vuotHanMuc, "don 49 trieu + 2 trieu -> vuot han muc 50 trieu");
    // Bo dung co gia tri mac dinh cho MOI thu, va test chi noi ra thu no QUAN TAM.

    // ---- 3. BAT BIEN: kiem voi NGHIN dau vao, khong phai ba ----
    std::mt19937 rnd(42);                         // hat giong CO DINH -> tai hien duoc
    std::uniform_int_distribution<long> dGia(1, 1000000);
    std::uniform_int_distribution<int> dSl(1, 40), dNguoi(1, 9);
    int soCaChay = 0;
    for (int i = 0; i < 1000; ++i) {
        KhoGia k;
        DonHang d = DonHangBuilder().voiDong("x", dGia(rnd), dSl(rnd)).dung(k);
        int nguoi = dNguoi(rnd);
        long tong = d.tongTien(), s = 0, lon = 0, nho = tong;
        for (int j = 0; j < nguoi; ++j) {
            long p = phanThuI(tong, nguoi, j);
            s += p;
            if (p > lon) lon = p;
            if (p < nho) nho = p;
        }
        check(s == tong, "BAT BIEN: tong cac phan = tong ban dau");
        check(lon - nho <= 1, "BAT BIEN: chenh lech giua cac phan khong qua 1 don vi");
        ++soCaChay;
    }
    check(soCaChay == 1000, "1.000 ca sinh ngau nhien, 0 dong du lieu go tay");
    // Hat giong co dinh la bat buoc: mot test do ngau nhien ma khong tai hien duoc thi vo
    // dung — va te hon, no se bi danh dau "bo qua".

    // ---- 4. CON BUG: BAN GIA NOI DOI ----
    KhoGia khoNoiDoi;
    khoNoiDoi.giuCho("laptop", -5);               // so luong AM
    check(khoNoiDoi.daGiu["laptop"] == -5, "ban gia nhan so luong am — test XANH");

    bool banThatNo = false;
    try { KhoThat().giuCho("laptop", -5); }
    catch (const std::invalid_argument&) { banThatNo = true; }
    check(banThatNo, "ban that nem ngoai le — bug di thang ra production");
    // Cach chua la BO KIEM TRA HOP DONG (bai 68): mot bo test viet mot lan, chay tren MOI
    // cai dat cua cong. Ban gia nao khong qua duoc thi khong duoc dung.
    KhoGia g;
    KhoThat t;
    std::vector<KhoHang*> moiCaiDat{&g, &t};
    int soCaiDatQua = 0;
    for (KhoHang* k : moiCaiDat) {
        bool chan = false;
        try { k->giuCho("x", -1); } catch (const std::invalid_argument&) { chan = true; }
        if (chan) ++soCaiDatQua;
    }
    check(soCaiDatQua == 1, "chay hop dong tren 2 cai dat -> lo ra ngay cai nao noi doi");

    // ---- 5. DIEU CHI C++ CO: TEST CHAY LUC BIEN DICH ----
    // Nam `static_assert` o dau file da chay XONG truoc khi chuong trinh nay ton tai.
    // Chung kiem dung ba bat bien o phan 3, nhung:
    //   - ton 0 mili-giay luc chay;
    //   - khong the bi "bo qua", khong the bi tat bang co dong lenh;
    //   - code sai thi KHONG TAO RA DUOC file thuc thi.
    // Do la vong phan hoi nhanh nhat co the: khong phai "test do sau 3 giay" ma la "build
    // do ngay trong trinh soan thao".
    //
    // Gioi han — phan nao cua mien dua vao day duoc:
    //   DUOC   : phep tinh thuan tren so/kieu don gian (chia tien, chuyen doi don vi,
    //            bang chuyen trang thai — bai 89 phan 5).
    //   KHONG  : bat cu thu gi dung `std::string` phuc tap, cap phat dong, hay ngoai le.
    // Nen `static_assert` KHONG thay the duoc bo test — no phu them mot lop, cho dung
    // nhung bat bien thuan tuy nhat va quan trong nhat.
    constexpr long kiemLucBienDich = tongPhan(1700000, 3);
    static_assert(kiemLucBienDich == 1700000, "bat bien cua don hang o phan 1");
    check(kiemLucBienDich == 1700000, "va van doc duoc luc chay nhu mot hang so");

    // ---- 6. TEST MIEN KHONG CAN KHUNG PHAN MEM ----
    // Toan bo file nay — va moi file self-check trong series — la ham + `check`. Khong
    // Google Test, khong mock, khong fixture, 0 mili-giay khoi dong.
    //
    // Do khong phai "luoi khong dung Google Test". No la BANG CHUNG rang mo hinh mien da
    // tach sach khoi ha tang (bai 98). Phep thu nguoc lai cung dung:
    //
    //   NEU TEST MIEN CUA BAN CAN MOT KHUNG PHAN MEM DE CHAY,
    //   THI THU BAN DANG TEST KHONG PHAI MIEN.
    check(donA.ma() == "DH-01", "mot ham, mot check, khong ha tang nao");

    // ---- 7. CAI GI KHONG NEN TEST, VA DAT TEN THE NAO ----
    //   - Getter thuan: `check(d.ma() == "DH-01")` khong kiem duoc luat nao.
    //   - Thu vien chuan: `std::map` da duoc test roi.
    //   - Cai dat rieng tu: neu phai doi `private` thanh `public` de test, thi test do
    //     dang bam vao cach lam (phan 1) — hay test qua cua chinh.
    // Nguoc lai, thu DANG test nhat la nhung cho co `if` mang nghia nghiep vu: han muc,
    // chuyen trang thai (bai 89), luat gia, luat chia tien.
    //
    // Va ten test la mot cau nghiep vu:
    //   TE : "test1", "testThemDong"
    //   TOT: "don 49 trieu + 2 trieu -> vuot han muc 50 trieu"
    // Khi test do luc 2 gio sang, dong chu do la toan bo thu nguoi truc co.
    check(DonHang::HAN_MUC == 50000000L, "luat nghiep vu co `if` -> dang test");

    std::cout << "OK\n";
    return 0;
}
